File: [0338][Medium][Counting_Bits]/Counting_Bits_4.java
// leetcode time     cost : 1 ms
// leetcode memory   cost : 43.1 MB 
// Time  Complexity: O(n)
// Space Complexity: O(n)
// slolution 4, DP and lowest set bit,
// P(i)=P( i&(i-1) ) + 1, use x &= x - 1 to get the lowest set bit,
public class Solution {
    public int[] countBits(int num) {
        int[] ans = new int[num + 1];
        for (int i = 1; i <= num; ++i)
          ans[i] = ans[i & (i - 1)] + 1;
        return ans;
    }
}

File: [0338][Medium][Counting_Bits]/Counting_Bits_4.py
# leetcode time     cost : 44 ms
# leetcode memory   cost : 14.6 MB 
# Time  Complexity: O(n)
# Space Complexity: O(n)
# slolution 4, DP and lowest set bit,
# P(i)=P( i&(i-1) ) + 1, use x &= x - 1 to get the lowest set bit,
class Solution:
    def countBits(self, num: int) -> List[int]:
        result = [0 for i in range(num+1)]
        for num_i in range(1,num+1):
            result[num_i] = result[num_i & (num_i - 1)] + 1
        return result